dig/dig.go

```diff
@@ -30,12 +30,39 @@ import (
 )
 
 var (
-	errParamType   = errors.New("graph injection must be done through a pointer or a function")
-	errReturnCount = errors.New("constructor function must return exactly one value")
-	errReturnKind  = errors.New("constructor return type must be a pointer")
-	errArgKind     = errors.New("constructor arguments must be pointers")
+	errParamType     = errors.New("graph injection must be done through a pointer or a function")
+	errReturnCount   = errors.New("constructor function must return up to two values")
+	errReturnKind    = errors.New("constructor return type must be a pointer")
+	errReturnErrKind = errors.New("second return value of constructor must be error")
+	errArgKind       = errors.New("constructor arguments must be pointers")
+
+	_typeOfError = reflect.TypeOf((*error)(nil)).Elem()
 )
 
+// Wraps a function that returns a single value to return that value and a nil
+// error.
+func addErrorReturn(c interface{}) interface{} {
+	v := reflect.ValueOf(c)
+	t := v.Type()
+
+	if t.NumOut() != 1 {
+		panic(fmt.Sprintf("addErrorReturn called with invalid function: %v", t))
+	}
+
+	returnTypes := []reflect.Type{t.Out(0), _typeOfError}
+	argTypes := make([]reflect.Type, t.NumIn())
+	for i := 0; i < t.NumIn(); i++ {
+		argTypes[i] = t.In(i)
+	}
+
+	return reflect.MakeFunc(
+		reflect.FuncOf(argTypes, returnTypes, t.IsVariadic()),
+		func(args []reflect.Value) []reflect.Value {
+			return append(v.Call(args), reflect.Zero(_typeOfError))
```

Comment:
Validation should probably happen in funcNode instead. This function was deleted based on @glibsm's comment so this isn't necessary anyway.